### PR TITLE
Use `thiserror` for error type

### PR DESCRIPTION
**Pull request by Spaceface16518**
Details of implementation in commit message and commit comments.

# Library

[`thiserror`](https://lib.rs/crates/thiserror) is an error handling helper library that uses procedural macros to derive `std::error::Error` and `Display`, as well as any needed `From` impls.

# Benefits

This approach is beneficial because error descriptions and variable interpolations are located directly next to enum variant definitions, making the code more readable and understandable.

# Limitations

This cleaner code comes at the price of compile time. On my machine (`AMD Ryzen 7 4700U with Radeon Graphics (8) @ 2.000GHz`, 8GB memory), this branch takes about 4 seconds longer to compile than the master branch.

| Command | Mean [s] | Min [s] | Max [s] |
|:---|---:|---:|---:|
| `master` | 21.120 ± 0.263 | 20.750 | 21.630 |
| `thiserror` | 25.156 ± 0.306 | 24.454 | 25.576 |

# Licensing

`thiserror` [uses the MIT License](https://github.com/dtolnay/thiserror/blob/master/LICENSE-MIT), so it should be compatible with this repository's license. It can also licensed under Apache, if that's easier. I'm not super familiar with licensing, so I'm not sure if this is a challenge.

### CODE REVIEW

**Comment by Spaceface16518** [comment]
I can't tell if [this test failure](https://github.com/signalapp/libsignal-client/pull/140/checks?check_run_id=1683898582) is being caused by my changes or if it's a known issue in the CI right now. Is it possible `thiserror` does not support `armv7-linux-androideabi`?

**Comment by jrose-signal** [comment]
Known issue. Jack's taking care of it in #139.

**Comment by Spaceface16518** [comment]
Thank you @jrose-signal.

File: rust/aes-gcm-siv/src/error.rs [hunk 1]
```diff
@@ -3,27 +3,16 @@
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
-use std::fmt;
-
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(thiserror::Error, Clone, Debug, Eq, PartialEq)]
```
Comment (Spaceface16518) [inline rust/aes-gcm-siv/src/error.rs:6]: This causes `Error` to implement `std::error::Error`. This _is_ a behavioral change, as this error type did not implement the std error trait before this commit.

**Review by Spaceface16518 — COMMENTED** [review]
This satisfies [the clippy lint](https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call) that made the [previous CI run](https://github.com/signalapp/libsignal-client/runs/1684228606?check_suite_focus=true) fail.

**Comment by jrose-signal** [comment]
I just merged #114, which means some reworking needed of this patch, but I want to warn you that we probably won't get to this even as a pure refactor until next week or the week after due to [all the attention Signal is getting right now](https://twitter.com/signalapp/status/1347693761044701186). Please don't take it personally!

**Comment by Spaceface16518** [comment]
> I just merged #114, which means some reworking needed of this patch

no problem, i've got time on my hands 😁 

> I want to warn you that we probably won't get to this even as a pure refactor until next week or the week after…Please don't take it personally!

no offense to taken, i completely understand!

> …due to [all the attention Signal is getting right now](https://twitter.com/signalapp/status/1347693761044701186).

This is incredible! i'm so excited for this project and so happy to be contributing to it.

**Comment by Spaceface16518** [comment]
Okay, I just spent multiple hours on a rebase, and it's still saying there are merge conflicts. Is it okay if I just use a merge commit to resolve the changes, or should I keep attempting a rebase?

I apologize, I'm just not very experienced with git. I know it's a busy time right now, but any help/advice would be appreciated.

**Comment by hiqua** [comment]
@Spaceface16518 maybe you'd find it easier to start from the current `master` from scratch, and `git cherry-pick` all your commits and resolve their corresponding conflicts manually.

**Comment by Spaceface16518** [comment]
I think that will work! Thank you!

**Comment by jrose-signal** [comment]
Sorry for leaving this to sit—as you may have seen there's been a bunch of reorganization pulling common work into rust/bridge/shared/. I think I've just merged the last of the major disruptive changes, though!

**Comment by Spaceface16518** [comment]
@jrose-signal is it okay if I just scrap this PR and open a new one? I don't want to do too much force-pushing given (1) the level of commit-dependent conversation in this PR that would still apply to the next one and (2) the high-profile nature of this project.

**Comment by jrose-signal** [comment]
I think we aren't going to take this PR after all. The code that's being replaced may be a bit longer, but it's not really any more complicated than the `thiserror` attributes, and in cases like this with roughly even complexity we'd rather stick to standard Rust.

Sorry for taking so long to come to this decision, and please don't think it reflects poorly on you or the quality of the PR!

**Comment by Spaceface16518** [comment]
It's okay, I understand the reasoning. I'll look for other things I can contribute to the signal project. Thank you for your your friendly and informative communication!

**Pull request closed without merging**